File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/js/JSShipment.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.js

import com.getcapacitor.JSObject
import com.microblink.core.Shipment
import org.json.JSONArray
import org.json.JSONObject

/**
 * Represents a Receipt Shipment for response purposes.
 *
 * This class encapsulates information about a shipment from a receipt, including its status
 * and the list of products in the shipment.
 *
 * @param shipment The [Shipment] object to be converted to an JSShipment.
 */
class JSShipment(shipment: Shipment)  {
    private val status: String
    private val products: List<JSProduct>

    /**
     * Initializes an JSShipment instance based on the provided [shipment].
     *
     * @param shipment The [Shipment] object to be converted to an JSShipment.
     */
    init {
        status = shipment.status()
        products = shipment.products().map { product -> JSProduct(product) }
    }

    /**
     * Converts the JSShipment object to a JSON representation.
     *
     * @return A [JSONObject] containing the status and products of the shipment.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("status", status)
            .put("products", JSONArray(products.map { prd -> prd.toJS() }))

    companion object {
        /**
         * Creates an JSShipment object if the provided [shipment] is not null.
         *
         * @param shipment The [Shipment] object to be converted to an JSShipment.
         * @return An [JSShipment] object if [shipment] is not null; otherwise, null.
         */
        fun opt(shipment: Shipment?): JSShipment? =
            if (shipment != null) JSShipment(shipment) else null
    }
}

File: ios/Plugin/Plugin/Js/JSShipment.swift
/*
 * RspShipment Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import BlinkReceipt
import BlinkEReceipt
import Capacitor

/**
 Represents a response containing shipment information.

 This struct is used to convey details about a shipment, including its status and the list of products within the shipment.
 */
struct JSShipment {
    /// The status of the shipment.
    private let status: String
    /// The list of products within the shipment.
    private let products: [JSProduct]

    /**
     Initializes an `RspShipment` struct.

     - Parameters:
        - shipment: The shipment information.
     */
    init(shipment: BRShipment) {
        status = shipment.status
        products = shipment.products?.map { product in JSProduct(product: product) } ?? []
    }
    
    static func opt(shipment: BRShipment?) -> JSShipment? {
        return shipment != nil ? JSShipment(shipment: shipment!) : nil
    }

    /**
     Converts the `RspShipment` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing shipment information in a format suitable for a Capacitor plugin call result.
     */
    func toJSObject() -> JSObject {
        var ret = JSObject()
        ret["status"] = status
        ret["products"] = JSArray(arrayLiteral: products.map { prd in prd.toJSObject() })
        return ret
    }
}

